### PR TITLE
refactor net_ebpf_ext

### PR DESCRIPTION
**Pull request by shankarseal**
This PR is a simple refactoring of net_ebpf_ext that separates bind and xdp hooks into their own source files and performs some cleanups. This is in preparation for #341 .

### CODE REVIEW

File: netebpfext/net_ebpf_ext.h [hunk 1]
```diff
@@ -18,8 +18,30 @@
...
+#include <netiodef.h>
+#include <ntddk.h>
+
+#include "ebpf_ext_attach_provider.h"
+#include "ebpf_nethooks.h"
+#include "ebpf_platform.h"
+#include "ebpf_program_types.h"
+#include "ebpf_windows.h"
+
+#define RTL_COUNT_OF(arr) (sizeof(arr) / sizeof(arr[0]))
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.h:39]: We already have
```
libs\platform\ebpf_platform.h:#define EBPF_COUNT_OF(arr) (sizeof(arr) / sizeof(arr[0]))
```

Can't we just use that and delete this line?

File: netebpfext/net_ebpf_ext_bind.c [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+Abstract:
+
+   This file implements the BIND program type hook on eBPF for Windows.
+
+Environment:
+
+    Kernel mode
+
+--*/
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_bind.c:14]: Replace this with doxygen style comment

File: netebpfext/net_ebpf_ext_bind.c [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+/* ++
+
+   A simple classify function at the WFP Resource Allocation layer.
+
+-- */
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_bind.c:71]: replace with doxygen style comment.  Same for line 129 etc.  doxygen comments always go above the prototype, not below.

File: netebpfext/net_ebpf_ext_drv.c [hunk 4]
```diff
@@ -113,61 +116,43 @@ __net_ebpf_ext_driver_initialize_objects(
...
-    device_create_flag = TRUE;
+    _net_ebpf_ext_driver_device_object = WdfDeviceWdmGetDeviceObject(_net_ebpf_ext_device);
 
-    status = net_ebpf_ext_register_providers();
-    if (!NT_SUCCESS(status)) {
-        goto Exit;
-    }
+    net_ebpf_ext_register_callouts(_net_ebpf_ext_driver_device_object);
+    // ignore status. at boot, registration can fail.
+    // we will try to re-register during program load.
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_drv.c:123]:
```suggestion
    (void)net_ebpf_ext_register_callouts(_net_ebpf_ext_driver_device_object);
    // Ignore status. At boot, registration can fail.
    // We will try to re-register during program load.
```

File: netebpfext/net_ebpf_ext_xdp.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+Abstract:
+
+   This file implements the XDP program type hook and helper functions on eBPF for Windows.
+
+Environment:
+
+    Kernel mode
+
+--*/
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_xdp.c:14]: Replace with doxygen style comment.

File: netebpfext/net_ebpf_ext_xdp.c [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+/* ++
+
+   A simple classify function at the WFP L2 MAC layer.
+
+-- */
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_xdp.c:99]: use doxygen style

File: netebpfext/net_ebpf_ext_drv.c (hunk 4 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_drv.c]: This comment is wrong. I have filed an issue.

File: netebpfext/net_ebpf_ext_drv.c [hunk 7]
```diff
@@ -32,58 +32,61 @@ WDF based driver that does the following:
...
 
 //
 // Pre-Declarations
 //
 DRIVER_INITIALIZE DriverEntry;
 
-static _Function_class_(EVT_WDF_DRIVER_UNLOAD) _IRQL_requires_same_
-    _IRQL_requires_max_(PASSIVE_LEVEL) void __net_ebpf_ext_driver_unload(_In_ WDFDRIVER driver_object)
+static void
+__net_ebpf_ext_driver_uninitialize_objects()
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_drv.c:45]:
```suggestion
_net_ebpf_ext_driver_uninitialize_objects()
```
Only one leading underscore.  Leading double underscores are reserved by the compiler.

https://en.cppreference.com/w/c/language/identifier#Reserved_identifiers
"Note: in C++, identifiers with a double underscore anywhere are reserved everywhere; in C, only the ones that begin with a double underscore are reserved."

File: netebpfext/net_ebpf_ext_drv.c [hunk 8]
```diff
@@ -32,58 +32,61 @@ WDF based driver that does the following:
...
     net_ebpf_ext_unregister_providers();
 
     net_ebpf_ext_unregister_callouts();
+
+    if (_net_ebpf_ext_device != NULL)
+        WdfObjectDelete(_net_ebpf_ext_device);
+}
+
+static _Function_class_(EVT_WDF_DRIVER_UNLOAD) _IRQL_requires_same_
+    _IRQL_requires_max_(PASSIVE_LEVEL) void __net_ebpf_ext_driver_unload(_In_ WDFDRIVER driver_object)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_drv.c:58]:
```suggestion
    _IRQL_requires_max_(PASSIVE_LEVEL) void _net_ebpf_ext_driver_unload(_In_ WDFDRIVER driver_object)
```

File: netebpfext/net_ebpf_ext_drv.c [hunk 9]
```diff
@@ -32,58 +32,61 @@ WDF based driver that does the following:
...
+// Create and initialize WDF driver, device object,
+// WFP callouts and NPI providers.
 //
 static NTSTATUS
-__net_ebpf_ext_driver_initialize_objects(
-    _Inout_ DRIVER_OBJECT* driver_object,
-    _In_ const UNICODE_STRING* registry_path,
-    _Out_ WDFDRIVER* driver,
-    _Out_ WDFDEVICE* device)
+__net_ebpf_ext_driver_initialize_objects(_Inout_ DRIVER_OBJECT* driver_object, _In_ const UNICODE_STRING* registry_path)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_drv.c:69]:
```suggestion
_net_ebpf_ext_driver_initialize_objects(_Inout_ DRIVER_OBJECT* driver_object, _In_ const UNICODE_STRING* registry_path)
```

File: netebpfext/net_ebpf_ext_xdp.c [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+static ebpf_program_data_t _ebpf_xdp_program_data = {&_ebpf_xdp_program_info, NULL};
+
+static ebpf_extension_data_t _ebpf_xdp_program_info_provider_data = {
+    NET_EBPF_EXTENSION_NPI_PROVIDER_VERSION, sizeof(_ebpf_xdp_program_data), &_ebpf_xdp_program_data};
+
+// 85e0d8ef-579e-4931-b072-8ee226bb2e9d
+DEFINE_GUID(EBPF_ATTACH_TYPE_XDP, 0x85e0d8ef, 0x579e, 0x4931, 0xb0, 0x72, 0x8e, 0xe2, 0x26, 0xbb, 0x2e, 0x9d);
+
+DEFINE_GUID(EBPF_PROGRAM_TYPE_XDP, 0xf1832a85, 0x85d5, 0x45b0, 0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0);
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_xdp.c:39]: add guid comment above this line or delete line 36

**Review by dthaler — APPROVED** [review]
ok other than removing the double underscores

File: netebpfext/net_ebpf_ext_bind.c [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+void
+net_ebpf_ext_resource_allocation_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
```
Comment (Alan-Jowett) [inline netebpfext/net_ebpf_ext_bind.c:62]: Looks like this only handles IPv4? Do we need an issue tracker to add IPv6 support?

**Review by Alan-Jowett — APPROVED** [review]

File: netebpfext/net_ebpf_ext_bind.c (hunk 11 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_bind.c:62]: Yes #526 filed for this.

**Pull request merged**